### DISCUSSION TITLE
Running out of memory when clustering -- time to move on from DuckDB?

### DISCUSSION THREAD

**Discussion by ghost** [moj-analytical-services/splink#1390] [Q&A] [answered]
I'm doing some exploration of using Splink to match and dedupe company records across datasets. I'm not an engineer so I don't have a good feel for what's not working due to my lack of optimisation, and what's not working because I've hit the edges of what my platform can handle! I'd really appreciate some advice on whether I should push through this, or look to infrastructure for support.

I'm currently running everything on a Jupyter instance with about 30Gb memory using the DuckDB backend, pulling from a Postgres database. When I run predict I get a ~80m row dataframe that's about 11Gb in memory, but when I run cluster, it dumps a dozen 2Gb temp files on the drive then crashes the kernel. I've stepped through the connected components function and it's generating the initial neighbours and representatives tables that kills it. 

I'm experimenting with a few ways through:

* Iterative clustering using the assumptions in [this answer](https://github.com/moj-analytical-services/splink/discussions/1218)
* Using networkx to do the connected components (also kills memory right now)
* Replacing unique ID strings with integer representations

But I wonder if I'm just reaching the upper limit of what an in-memory DBMS is ever going to achieve. Am I missing something or is it time to start thinking about using Spark? Are there optimisations I'm missing that would make working with 80m rows feasible on an instance of that size?

**Comment by RobinL** [comment]
Hiya.  Good question.  The first to try (if you're not already doing it) is to use an on-disk duckdb database rather than the default in memory one
https://github.com/moj-analytical-services/splink/blame/4935a6a48e3fb32003687d748969b96ba3648ce9/splink/duckdb/linker.py#L125
https://duckdb.org/docs/archive/0.4.0/api/python.html#startup--shutdown

If that doesn't work there are some other options we can try

**Reply by ghost** [reply]
I was convinced you'd got it in one! Got into the iterations for the first time, looked like it converged, then crashed the kernel 😔

I'll step through in the morning and report back. Thank you so much for the speedy reply!

**Reply by RobinL** [reply, marked as answer]
Interesting - not what I would have expected, but we haven't tested much with on-disk duckdb dbs.  Specifically (like you) I'm surprised that an operation that worked in-memory would crash when attempting on-disk.

A few more specifics to try tomorrow:

1.  For data ingestion, ensure that you copy the data into the on-disk database file prior to using Splink.   DuckDB ingestion docs [here](https://duckdb.org/docs/api/python/data_ingestion).  i.e. avoid the route of e.g. loading the data into a pandas dataframe `df` and then just using `con.register('my_view', df).  

```python
con.execute('CREATE TABLE test_df_table AS SELECT * FROM test_df')
```

Once you've done this, you should see the filesize of your on-disk database (e.g. `my-db.duckdb`) increase.

To minimise memory usage, I recommend doing this as a separate script.  Once the data is ingested, restart the python kernel, and do the Splink work separately.  (Also worth checking to ensure you've closed all other python kernels in jupyter)

2.  You mentioned trying to shorten the unique ids.  That's a good idea.  If you're inputting multiple datasets, when you load the linker, there may be some mileage in also making their aliases as short as possible.  
```python
con = duckdb.connect(database='my-db.duckdb', read_only=True)
linker = DuckDBLinker(
    [df_left, df_right], settings, input_table_aliases=["l", "r"], connection=con
)
```

3.  You might want to consider turning your ipynbs into .py files and running them from the terminal.  You could for instance have an ingestion, a training, a predict, and a clustering.py.

4.  If all of the above fails, we have a new more-efficient approach to EM estimation that's just dropped in `master` on Github (it's not released yet).  You can see the PR [here](https://github.com/moj-analytical-services/splink/pull/1369)

**Reply by ghost** [reply]
It worked!! Summarising what I did for future users:

* Created duckdb database on disk
* Crated lookup table for aliases and unique IDs to integers from my cleaned, stored parquet extracts
* Copied all data into the disk database and replaced the unique IDs with their lookup integer
* Instantiated linker using the on-disk db, using the table lookup for names and integer aliases
* Ran linker as normal -- train, predict, cluster
* Joined lookups to cluster df and exported to parquet

The only wobble I had was a cluster crash in the middle of an early run. Because the predictions are in the physical database we can retrieve them, but it's a bit counterintuitive how to reconstruct the `SplinkDataFrame`. In the end I went with:

```python
predict_table = con.query("""
    select table_name
    from information_schema.tables
    where table_name like '%predict%';
""").fetchone()[0]
predictions = linker.register_table(predict_table, predict_table)
```

Bit of polishing to do but I think we've over the hump. Many, many thanks @RobinL !

**Reply by RobinL** [reply]
Excellent!  Thanks for the detailed write up!

**Reply by NickCrews** [reply]
@will-langdale-dit thanks for the writeup!

> Created duckdb database on disk

My impression was that if even if you used `:memory`, then duckdb would gracefully spill to disk if you started needing more memory than your machine has. Are you sure that changing away from `:memory` was actually important here?

If I understand https://github.com/duckdb/duckdb/issues/6733 correctly, some operations in duckdb such as hash joins require everything to fit into memory, so even if you use a `.duckdb` file then you still can run out of memory on hug joins. Do I have that right?

**Reply by ghost** [reply]
It could have been that the replacing labels with integers was more important. The thing was definitely spilling onto disk already as when it crashed I'd have a huge `.tmp/` dir to remove or run out of space on our cloud platform. I've not tested it, @NickCrews , but regardless I think it's better to make the decision actively rather than passively.

**Reply by NickCrews** [reply]
Yeah I bet the integer keys was more important as well. Good to know re spilling to disk. Thank you!
